### PR TITLE
Move-storage-to-persistence-layer

### PR DESCRIPTION
**Pull request by farmanahmed888**
- Added MongoDB integration in insti.py to fetch content from the database instead of a local JSON file.
- Created test_comprehension_structure.py to validate the hydration of audio fields in comprehension structures.
- Implemented test_insti.py to test the instantiation of FSM from the latest content using a dummy MongoDB class.
- Updated existing functions to accommodate the new MongoDB structure and ensure compatibility with the test cases.

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
PR Review Summary:

Overall, the modifications aim to address URL hydration logic, testing coverage, and some code cleanup. There are notable positive points, but multiple areas need improvement for maintainability, clarity, and robustness.

Strengths:
- The `ContentWebApp/db.json` removal cleanly replaces static paths with dynamic URL generation via `build_audio_url`.
- The introduction of hydrate logic in `comprehension_structure.py` is correctly designed to convert relative paths into full URLs at runtime, which makes content flexible and environment-agnostic.
- Test coverage in `test_comprehension_structure.py` and `test_insti.py` shows a focus on mocking external dependencies and verifying core functionality.
  
Main Concerns & Improvement Areas:
1. **Code Duplication & Inconsistencies:**
   - The old `db.json` is entirely removed, which is appropriate, but ensure no residual static content exists elsewhere.
   - The URL hydration logic in `comprehension_structure.py` appears solid, but be cautious that it relies on consistent keys (`titleAudioPath`, `themeAudioPath`). Enforce schema validation for robustness.
   
2. **Type Safety & Data Handling:**
   - The original JSON shows string booleans (`"true"/"false"`). The hydrate code presumes boolean types (`True/False`). This should be normalized, either during ingestion or in hydrate logic, to prevent runtime errors.
   - The test example uses fixed string booleans. Consider adding code to convert `"true"/"false"` strings to actual booleans before hydrate functions operating.
   
3. **Design & Complexity:**
   - The hydrate function and URL building are straightforward, but the overall content-loading pattern in `load_comprehension_structure` could benefit from explicit validation/normalization.
   - The test setup patches MongoDB and settings correctly; however, validation that the hydrate output exactly matches expected URLs is vital.
   - The `IVRv2/app/fsm/insti.py` test for async instantiation works but relies on minimal mockups. As complexity grows, consider more detailed scenario tests to ensure full content schema coverage.

4. **Code Quality & Readability:**
   - The `IVRv2/app/fsm/insti.py` has commented-out code and blocks that could be cleaned up or clarified.
   - Test functions could improve assertions, e.g., verify all URL parts, and explicitly test for missing/invalid keys to ensure error handling.
   
5. **Potential Enhancement (Not Critical Now):**
   - Consider centralizing URL schemes and hydration rules to avoid hardcoded paths scattered across different modules.
   - Validate content documents against a schema to prevent runtime errors, especially with inconsistent data types.

**Web Articles for Reference:**
- "Effective Python: Writing Cleaner, Safer Python Code": [https://realpython.com/effective-python-pythonic-code/](https://realpython.com/effective-python-pythonic-code/)
- "Data Normalization in Python": [https://realpython.com/python-data-normalization/](https://realpython.com/python-data-normalization/)
- "Mocking External Dependencies in Tests": [https://pytest-mock.readthedocs.io/en/latest/](https://pytest-mock.readthedocs.io/en/latest/)

**Score:** 73/100  
Despite solid approach, the code needs better data validation, removed duplication, and cleaner test assertions to reach a higher standard.

**Acceptance threshold met,** but with recommended improvements for long-term robustness and maintainability.

File: IVRv2/app/comprehension_structure.py [hunk 1]
```diff
@@ -1,236 +1,54 @@
...
+            if "audio_path" in opt:
+                opt["url"] = build_audio_url(opt["audio_path"])
+
+    return doc
+
+
+async def load_comprehension_structure():
+    """Load and hydrate comprehension structure from MongoDB."""
+    comprehension = MongoDB("comprehension")
+    db_doc = await comprehension.find_all()
```
Comment (Soumabha) [inline IVRv2/app/comprehension_structure.py]: A bit confusing. We seem to be processing as single document but find_all would technical return multiple rows.

We should either handle the list or use some 'find_by_one" types query.

File: IVRv2/app/fsm/insti.py [hunk 2]
```diff
@@ -459,14 +461,16 @@ def generate_states(fsm, content_list, content_attributes, level, parent_state_i
...
+    # # content = await get_content(content_ids)
+    # script_dir = os.path.dirname(os.path.abspath(__file__))  # Get script directory
+    # file_path = os.path.join(script_dir, "contentsV3.json")  # Build absolute path
 
-    print(file_path)
+    # print(file_path)
 
-    with open(file_path, "r", encoding="utf-8") as file:
-        content = json.load(file)
+    # with open(file_path, "r", encoding="utf-8") as file:
```
Comment (Soumabha) [inline IVRv2/app/fsm/insti.py]: remove lines not being used.

File: IVRv2/app/comprehension_structure.py [hunk 3]
```diff
@@ -1,236 +1,54 @@
...
+        for opt in q.get("options", []):
+            if "audio_path" in opt:
+                opt["url"] = build_audio_url(opt["audio_path"])
+
+    return doc
+
+
+async def load_comprehension_structure():
+    """Load and hydrate comprehension structure from MongoDB."""
+    comprehension = MongoDB("comprehension")
```
Comment (Soumabha) [inline IVRv2/app/comprehension_structure.py]: The database layer should not be loaded directly. It should be injected and used for querying.

File: IVRv2/app/fsm/insti.py
Comment (Soumabha) [inline IVRv2/app/fsm/insti.py:475]: What about this file? We need the visualization also to be dynamic from the database.

File: IVRv2/app/fsm/insti.py [hunk 4]
```diff
@@ -459,14 +461,16 @@ def generate_states(fsm, content_list, content_attributes, level, parent_state_i
 
 
 async def instantiate_from_latest_content(content_ids: Optional[List[str]] = None):
-    # content = await get_content(content_ids)
-    script_dir = os.path.dirname(os.path.abspath(__file__))  # Get script directory
-    file_path = os.path.join(script_dir, "contentsV3.json")  # Build absolute path
+    # # content = await get_content(content_ids)
```
Comment (Soumabha) [inline IVRv2/app/fsm/insti.py]: Remove any dead code to prevent confusion, esp for LLM supported coding.

File: IVRv2/app/comprehension_structure.py [hunk 5]
```diff
@@ -1,236 +1,54 @@
...
-      ],
-      "correct_option_id": "d94ce54d-09b5-432e-842d-e3a84ad62caa"
-    }
-  ]
-}
\ No newline at end of file
+output_container_path = "output-container/"
+
+
+def build_audio_url(relative_path: str) -> str:
```
Comment (Soumabha) [inline IVRv2/app/comprehension_structure.py]: Is this specific to the comprehension_structure? We should probably have it in a common azure storage extension class so that we can use the same logic across multiple files.

File: IVRv2/app/comprehension_structure.py [hunk 6]
```diff
@@ -1,236 +1,54 @@
...
+
+
+async def load_comprehension_structure():
+    """Load and hydrate comprehension structure from MongoDB."""
+    comprehension = MongoDB("comprehension")
+    db_doc = await comprehension.find_all()
+
+    if not db_doc:
+        raise ValueError("Comprehension document not found in MongoDB.")
+    print(f"Loaded comprehension structure from DB: {db_doc}")
```
Comment (Soumabha) [inline IVRv2/app/comprehension_structure.py]: We should use loggers rather than print statement.

File: IVRv2/app/comprehension_structure.py (hunk 5 above)
Comment (farmanahmed888) [reply IVRv2/app/comprehension_structure.py]: this is specific to comprehension_structure

File: IVRv2/app/comprehension_structure.py (hunk 3 above)
Comment (farmanahmed888) [reply IVRv2/app/comprehension_structure.py]: done

File: IVRv2/app/comprehension_structure.py (hunk 1 above)
Comment (farmanahmed888) [reply IVRv2/app/comprehension_structure.py]: done

**Review by a4i-architect — COMMENTED** [review]
PR Review Summary:

Overall, the codebase is undergoing a significant refactor toward more structured, testable, and maintainable components. The key change in the comprehension structure handling (app/utils/audio.py and app/comprehension_structure.py) demonstrates good practices by replacing hardcoded data with hydrated, URL-based logic, which is a positive evolution.

Strengths:
- Clear effort to modularize data hydration, especially converting relative paths to full URLs.
- Introduction of dependency injection patterns (app/dependencies.py) enhances testability.
- Improved interface abstraction (database.py) aligns with good design principles.
- Extensive test fixtures for comprehension data and mocks, aiding in robust testing.

Improvements & Concerns:
1. **Code Complexity & Cyclomatic Metrics:**
   - The original `IVRv2/app/fsm/instantiation.py` and `insti.py` files are heavily monolithic, with complex, nested logic, especially around state generation and transition management. This reduces readability and maintainability.
   - Refactoring these large functions into smaller, reusable functions with clear input/output could reduce cyclomatic complexity, e.g., state creation, URL assembly, and transition addition.

2. **Duplicate & Unnecessary Code:**
   - The JSON data files are lengthy and contain duplicate example entries (e.g., `contents.json`, `app/contents.json`), which should be condensed or retained only as fixtures in tests.
   - The comprehension document hydration functions in `audio.py` are on the right track but should be cleaner, avoiding hard-coded path strings scattered throughout.

3. **Type Safety & Typing:**
   - The database interface (`database.py`) now correctly uses asynchronous methods and type hints, which is good.
   - But methods like `find_by_id` or `update_document` return `Any` — typing could be more specific, e.g., `Optional[Dict[str, Any]]` for queries, `int` for counts, etc.

4. **Pattern & Architecture:**
   - The pattern of separating data access (repository) from core logic is present but could be more consistent, especially in `comprehension_repository.py`. It should standardize method signatures and include error handling.
   - The `load_comprehension_structure()` now correctly uses an async repository pattern; however, the function `hydrate_comprehension_document()` should ideally operate on the output of `get_all_comprehensions()` (a list), yet the comment suggests possible single-doc scenarios—clarify and enforce consistent data shapes.

5. **Testing & Mocks:**
   - The test `test_insti.py` uses patching to simulate MongoDB, which is good. However, the large data fixtures could be better structured as concise test data, avoiding overly verbose checks.
   - Cyclomatic complexity (~20+) suggests functions like `generate_states()` and nested state transitions are overly complicated. These could be split into smaller, deterministic functions with reduced branches.

6. **Design Patterns & Simplicity:**
   - The code is attempting to implement a finite-state machine (`FSM`) with states and transitions, which is apt. But large, intertwined functions lead to high cyclomatic complexity and should be more modular.
   - Adopting design patterns such as Command or Strategy for URL/Transition management could simplify logic.

7. **Additional Articles & Best Practices:**
   - [Reducing Cyclomatic Complexity](https://refactoring.guru/reduce-cyclomatic-complexity) — Break down large functions
   - [Clean Architecture in Python](https://medium.com/swlh/clean-architecture-in-python-part-1-2-68d0a89d8f86) — Clear separation of concerns
   - [Effective Use of Dependency Injection](https://martinfowler.com/articles/injection.html)

**Score: 72/100**  
While the code is on the right track with modularization and test coverage, overly complicated logic, inconsistent typings, and large monolithic functions prevent it from reaching higher standards. Focused refactoring of `generate_states()` and state management functions, standardization of data structures, and further type safety would elevate quality.

**Summary:**
- Good structure toward URL hydration and async data access.
- Needs breaking down complex functions and reducing cyclomatic complexity.
- Standardize data fixtures and ensure consistent, clear data schemas.
- Enhance type annotations and error handling.
- Implement more modular design patterns for FSM logic.

Acceptance is contingent on addressing these points, particularly simplifying control flows and improving test clarity.

**Review by a4i-architect — COMMENTED** [review]
Overall Review Summary:

The codebase has seen significant refactoring, especially around the comprehension structure and database interaction patterns. The primary improvements include moving data-related logic into async functions, better use of dependency injection, and a clear separation between data hydration and retrieval. The introduction of an interface for storage and database access (IDatabase) enhances testability.

Strengths:
- Transition from static JSON data to dynamic database fetching via an abstraction layer.
- Proper use of async functions to prevent blocking.
- Implementation of hydration functions that convert relative to full URLs, improving data consistency.
- Good use of dependency injection with caching (@lru_cache), especially for database services.
- Clear test fixtures, including mocks, which facilitate unit testing.

However, there are areas to improve:

Unnecessary/Ill-typed/Duplicates:
- The `db.json` file with all content is now stripped in favor of dynamic loading, which is good. But the JSON hints at redundant content, and the hydration functions could be optimized to avoid repeatedly adding full URLs.
- The `hydrate_comprehension_document` function is correctly separating concerns, but it relies on the presence of certain keys like `"titleAudioPath"`, which could be inconsistent. Defensive coding or schema validation would be better.
- The comprehension repository methods are minimal but functional. Consider adding explicit error handling if the collection or data is malformed.
- The `app/utils/audio.py` creates full URLs efficiently, but the `hydrate_comprehension_document` mutates input data; converting to immutable approaches might be cleaner to prevent side effects.
- The heavily complex commented-out code in `instantiation.py` suggests over-complicated logic that can be simplified with clear state machines and fewer nested conditions.
- Cyclomatic complexity in the commented code is high; refactoring into smaller, reusable functions or state pattern objects is recommended.

Design & Complexity:
- The state machine code (instantiation.py) appears overly complicated, with many nested conditionals, comment-heavy blocks, and procedural flows. A cleaner, object-oriented approach with explicit states/transitions should be considered.
- Use of string concatenation vs. `format()` is inconsistent. Prefer `f-strings` always for better readability.
- Data hydration is repetitive; leveraging schema validation (e.g., pydantic models) could improve robustness.

Testing:
- The test coverage for comprehension documents and database mocks is solid. The tests could be extended to cover validation errors and edge cases.
- The mock database implementation is good, but adding methods to simulate errors (e.g., connection issues) would be valuable.

Articles/Web References:
- Effective use of dependency injection with `@lru_cache`: https://realpython.com/dependency-injection-in-python/
- Hydrating relative URLs dynamically: https://python.readthedocs.io/en/stable/library/urllib.parse.html
- Managing complexity with state machines: https://refactoring.guru/design-patterns/state
- Testing async database access: https://pytest-asyncio.readthedocs.io/en/latest/

Score: 72/100

Summary:
The refactoring is generally on the right track, moving towards a more modular, testable, and scalable design. Yet, the complexity of the old state machine code remains, and some patterns (like mutation-heavy hydrations) could be better optimized or simplified. Focus on streamlining state management, confirming code consistency, and adding error handling will strengthen this codebase further.

Recommendations:
- Refactor the large, nested branch logic in `instantiation.py` into a dedicated state machine pattern.
- Use pydantic or similar validation to ensure consistency of loaded comprehension documents.
- Remove or update commented legacy code to reduce cognitive load.
- Extend tests to simulate error scenarios to improve robustness.

This will elevate code quality, maintainability, and scalability for future development.

**Review by Soumabha — APPROVED** [review]
